src/bsoncxx/builder/list/list.hpp

```diff
@@ -0,0 +1,112 @@
+// Copyright 2020 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <bsoncxx/builder/core.hpp>
+#include <bsoncxx/config/prelude.hpp>
+#include <bsoncxx/exception/error_code.hpp>
+#include <bsoncxx/exception/exception.hpp>
+#include <bsoncxx/types/bson_value/value.hpp>
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+namespace builder {
+using namespace bsoncxx::types;
+
+class list {
+    using initializer_list_t = std::initializer_list<list>;
+
+   public:
+    list() = default;
+
+    template <typename T>
+    list(T value) : _value{value} {}
+
+    list(initializer_list_t init, bool type_deduction = true, bool is_array = true) {
```

Comment:
This is really cool. It seems like reasonable default behavior to do type deduction.

Given that there are explicit ways to specify a `document` or `array` type with the nested classes (per this conversation: https://github.com/mongodb/mongo-cxx-driver/pull/739#discussion_r517448234) can we make this constructor private and have one public constructor that just takes an initializer list?

E.g.
```
public:
    list(initializer_list_t init) : list (init, true /* type_deduction */, true /* is_array */);

private:
    list(initializer_list_t init, bool type_deduction, bool is_array);
```

That hides the `type_deduction` and `is_array` arguments from the public API. If users want to opt out of type deduction, they do so explicitly with the nested classes.